trino/client.py

```diff
@@ -925,3 +977,217 @@ def _parse_retry_after_header(retry_after):
+
+
+# Trino Spooled protocol transfer objects
+class _SpooledProtocolResponseTO(TypedDict):
+    encoding: Literal["json", "json+std", "json+lz4"]
+    segments: List[_SegmentTO]
+
+
+class _SegmentMetadataTO(TypedDict):
+    uncompressedSize: str
+
+
+class _SegmentTO(_SegmentMetadataTO):
+    type: Literal["spooled", "inline"]
+    metadata: _SegmentMetadataTO
+
+
+class _SpooledSegmentTO(_SegmentTO):
+    uri: str
+    ackUri: str
+    headers: Dict[str, List[str]]
+
+
+class _InlineSegmentTO(_SegmentTO):
+    data: str
+
+
+class Segment:
+    """
+    Abstract base class representing a segment of data returned by a Trino query.
+
+    Attributes:
+        metadata (property): Metadata associated with the segment.
+        rows (property): Returns the decoded and mapped rows of data.
+    """
+    def __init__(self, segment: _SegmentTO) -> None:
+        self._segment = segment
+        self.metadata = segment["metadata"]
+
+    @property
+    @abstractmethod
+    def rows(self):
+        pass
+
+
+class InlineSegment(Segment):
+    """
+    A subclass of Segment that handles inline data segments. The data is base64 encoded and
+    requires mapping to rows using the provided row_mapper.
+
+    Attributes:
+        rows (property): The rows of data in the segment, decoded and mapped from the base64 encoded data.
+    """
+    def __init__(self, segment: _InlineSegmentTO, row_mapper: RowMapper) -> None:
+        super().__init__(segment)
+        self._segment = cast(_InlineSegmentTO, segment)
+        self._row_mapper = row_mapper
+
+    @property
+    def rows(self) -> List[List[Any]]:
+        inline_segment = self._segment
+        data = inline_segment["data"]
+        decoded_string = base64.b64decode(data)
+        rows = self._row_mapper.map(json.loads(decoded_string))
+        return list(iter(rows))
+
+    def __repr__(self):
+        return f"InlineSegment(metadata={self.metadata})"
+
+
+class SpooledSegment(Segment):
+    """
+    A subclass of Segment that handles spooled data segments, where data may be compressed and needs to be
+    retrieved via HTTP requests. The segment includes methods for acknowledging processing and loading the
+    segment from remote storage.
+
+    Attributes:
+        rows (property): The rows of data, loaded and mapped from the spooled segment.
+        uri (property): The URI for the spooled segment.
+        ack_uri (property): The URI for acknowledging the processing of the spooled segment.
+        headers (property): The headers associated with the spooled segment.
+
+    Methods:
+        acknowledge(): Sends an acknowledgment request for the segment.
+    """
+    def __init__(
+        self,
+        response: _SpooledProtocolResponseTO,
+        segment: _SpooledSegmentTO,
+        row_mapper: RowMapper,
+        request: TrinoRequest
+    ) -> None:
+        super().__init__(segment)
+        self._segment = cast(_SpooledSegmentTO, segment)
+        self._row_mapper = row_mapper
```

Comment:
we should probably construct row_mapper for a given encoding

Reply:
See above.